File: examples/pidinfo.c
/* Compile on macOS as:
   cc pidinfo.c -o pidinfo

   ./pidinfo <PID> will print out the process status
*/
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>

#include <libproc.h>

/* Status values from proc.h */
#define SIDL    1               /* Process being created by fork. */
#define SRUN    2               /* Currently runnable. */
#define SSLEEP  3               /* Sleeping on an address. */
#define SSTOP   4               /* Process debugging or suspension. */
#define SZOMB   5               /* Awaiting collection by parent. */

void print_status(struct proc_bsdshortinfo* proc) {
    char* status;
    switch (proc->pbsi_status) {
    case SIDL:
        status = "IDLE";
        break;
    case SRUN:
        status = "RUN";
        break;
    case SSLEEP:
        status = "SLEEP";
        break;
    case SSTOP:
        status = "STOP";
        break;
    case 5:
        status = "ZOMB";
        break;
    }
    printf(" status: %s\n", status);
    return;
}

int main(int argc, char *argv[])
{
    pid_t pid;
    struct proc_bsdshortinfo proc;

    if (argc == 2)
        pid = atoi(argv[1]);
    else
        pid = getpid();
    int st = proc_pidinfo(pid, PROC_PIDT_SHORTBSDINFO, 0,
                          &proc, PROC_PIDT_SHORTBSDINFO_SIZE);

    if (st != PROC_PIDT_SHORTBSDINFO_SIZE) {
        fprintf(stderr, "Cannot get process info");
        return 1;
    }
    printf(" pid: %d\n", (int)proc.pbsi_pid);
    printf("ppid: %d\n", (int)proc.pbsi_ppid);
    printf(" uid: %d\n", (int)proc.pbsi_uid);
    printf(" gid: %d\n", (int)proc.pbsi_gid);
    print_status(&proc);

    return 0;
}

File: examples/pidinfo.ml
(* Equivalent code to pidinfo.c this will print out the process status
   by calling libproc.h functions.

   dune exec -- pidinfo <PID>
 *)

let handle_call pid =
  let open Mach in
  let open Ctypes in

  (* Using ctypes allocate a bsd_shortinfo structure to pass into proc_pidinfo *)
  let bsd_shortinfo = allocate proc_bsdshortinfo (make proc_bsdshortinfo) in
  let i = proc_pidinfo pid 13 (* PROC_PIDT_SHORTBSDINFO *)
    (Unsigned.UInt64.zero)
    (to_voidp bsd_shortinfo)
    (sizeof proc_bsdshortinfo) in

  (* Check return size as this might have failed *)
  if i != (sizeof proc_bsdshortinfo) then
    failwith (Printf.sprintf "Unable to get SHORTBSDINFO returned: %d" i)
  else
    let status = getf (!@bsd_shortinfo) pbsi_status in
    Printf.printf "  pid: %u\n ppid: %d\n  uid: %d\n  gid: %d\n status: %s\n"
      (getf (!@bsd_shortinfo) pbsi_pid |> Unsigned.UInt32.to_int)
      (getf (!@bsd_shortinfo) pbsi_ppid |> Unsigned.UInt32.to_int)
      (getf (!@bsd_shortinfo) pbsi_uid |> Unsigned.UInt32.to_int)
      (getf (!@bsd_shortinfo) pbsi_gid |> Unsigned.UInt32.to_int)
      (pbi_status_of_int status |> Option.get |> pbi_status_to_string)

let () =
  if Array.length Sys.argv <> 2 then
    Printf.fprintf stderr "Usage: %s <pid>\n" Sys.executable_name
    else
      ignore (handle_call (PosixTypes.Pid.of_string Sys.argv.(1)))

